[5.3 regression] Failure when using the C API in C++ code

Trying to compile `mccs` with the `5.3` branch (https://github.com/ocaml/ocaml/commit/a4e7cbdb54204979294dfd0c0f801d342939246d) breaks with the following error message at least on macOS:
```
File "src_ext/mccs/src/dune", line 5, characters 63-73:
  5 |    new_criteria notuptodate_criteria removed_criteria mccscudf mccs_stubs)
                                                                     ^^^^^^^^^^
  (cd _build/default/src_ext/mccs/src && /usr/bin/gcc -O2 -fno-strict-aliasing -fwrapv -pthread -Wall -Wextra -Wno-unused-parameter -x c++ -DUSEGLPK -g -I /Users/runner/.cache/ocaml-local/lib/ocaml -I ../../cudf/lib -I ../../extlib/src -I glpk -o mccs_stubs.o -c mccs_stubs.cpp)
  In file included from mccs_stubs.cpp:7:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/mlvalues.h:20:
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/misc.h:353:1: error: expected expression
  CAMLnoret CAMLextern void caml_fatal_error (char *, ...)
  ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/misc.h:105:22: note: expanded from macro 'CAMLnoret'
    #define CAMLnoret [[noreturn]]
                       ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/misc.h:353:11: error: expected unqualified-id
  CAMLnoret CAMLextern void caml_fatal_error (char *, ...)
            ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/misc.h:140:34: note: expanded from macro 'CAMLextern'
  #define CAMLextern CAMLDLLIMPORT extern
                                   ^
  In file included from mccs_stubs.cpp:7:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/mlvalues.h:70:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:32:
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:17:1: error: expected parameter declarator
  DOMAIN_STATE(atomic_uintnat, young_limit)
  ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:44: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                             ^
  In file included from mccs_stubs.cpp:7:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/mlvalues.h:70:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:32:
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:17:1: error: expected ')'
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:44: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                             ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:17:1: note: to match this '('
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:34: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                   ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/misc.h:156:29: note: expanded from macro 'CAMLalign'
  #define CAMLalign(n) alignas(n)
                              ^
  In file included from mccs_stubs.cpp:7:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/mlvalues.h:70:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:32:
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:17:1: error: a type specifier is required for all declarations
  DOMAIN_STATE(atomic_uintnat, young_limit)
  ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:34: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                   ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/misc.h:156:22: note: expanded from macro 'CAMLalign'
  #define CAMLalign(n) alignas(n)
                       ^
  In file included from mccs_stubs.cpp:7:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/mlvalues.h:70:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:32:
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:17:14: error: expected ';' at end of declaration list
  DOMAIN_STATE(atomic_uintnat, young_limit)
               ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:28:1: error: expected parameter declarator
  DOMAIN_STATE(value*, young_ptr)
  ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:44: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                             ^
  In file included from mccs_stubs.cpp:7:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/mlvalues.h:70:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:32:
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:28:1: error: expected ')'
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:44: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                             ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:28:1: note: to match this '('
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:34: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                   ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/misc.h:156:29: note: expanded from macro 'CAMLalign'
  #define CAMLalign(n) alignas(n)
                              ^
  In file included from mccs_stubs.cpp:7:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/mlvalues.h:70:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:32:
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:28:1: error: a type specifier is required for all declarations
  DOMAIN_STATE(value*, young_ptr)
  ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:34: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                   ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/misc.h:156:22: note: expanded from macro 'CAMLalign'
  #define CAMLalign(n) alignas(n)
                       ^
  In file included from mccs_stubs.cpp:7:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/mlvalues.h:70:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:32:
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:28:14: error: expected ';' at end of declaration list
  DOMAIN_STATE(value*, young_ptr)
               ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:31:1: error: expected parameter declarator
  DOMAIN_STATE(value*, young_start)
  ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:44: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                             ^
  In file included from mccs_stubs.cpp:7:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/mlvalues.h:70:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:32:
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:31:1: error: expected ')'
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:44: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                             ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:31:1: note: to match this '('
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:34: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                   ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/misc.h:156:29: note: expanded from macro 'CAMLalign'
  #define CAMLalign(n) alignas(n)
                              ^
  In file included from mccs_stubs.cpp:7:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/mlvalues.h:70:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:32:
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:31:1: error: a type specifier is required for all declarations
  DOMAIN_STATE(value*, young_start)
  ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:34: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                   ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/misc.h:156:22: note: expanded from macro 'CAMLalign'
  #define CAMLalign(n) alignas(n)
                       ^
  In file included from mccs_stubs.cpp:7:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/mlvalues.h:70:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:32:
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:31:14: error: expected ';' at end of declaration list
  DOMAIN_STATE(value*, young_start)
               ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:34:1: error: expected parameter declarator
  DOMAIN_STATE(value*, young_end)
  ^
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:44: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                             ^
  In file included from mccs_stubs.cpp:7:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/mlvalues.h:70:
  In file included from /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:32:
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.tbl:34:1: error: expected ')'
  /Users/runner/.cache/ocaml-local/lib/ocaml/caml/domain_state.h:31:44: note: expanded from macro 'DOMAIN_STATE'
  #define DOMAIN_STATE(type, name) CAMLalign(8) type name;
                                             ^
[...]
```
After talking with @Octachron i'll try to go back through https://github.com/ocaml/ocaml/pull/11557 soon and polishing it in hope this would fix this kind of error forever

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -14,15 +14,11 @@ jobs:
           - macos-latest
           - windows-latest
         ocaml-compiler:
-          - 4.12.x
-          - 4.13.x
-          - 4.14.x
-          - 5.00.x
-        exclude:
-          - os: windows-latest
-            ocaml-compiler: 5.00.x
-          - os: macos-latest
-            ocaml-compiler: 4.08.x
+          - "4.14"
+          - "5.0"
+          - "5.1"
+          - "5.2"
+          - "5.3"
 
     runs-on: ${{ matrix.os }}
 
@@ -31,7 +27,7 @@ jobs:
         uses: actions/checkout@v3
 
       - name: Use OCaml ${{ matrix.ocaml-compiler }}
-        uses: ocaml/setup-ocaml@v2
+        uses: ocaml/setup-ocaml@v3
         with:
           ocaml-compiler: ${{ matrix.ocaml-compiler }}
 

--- a/src/context_flags.ml
+++ b/src/context_flags.ml
@@ -23,7 +23,7 @@ let cxxflags =
   let flags =
     (if (Sys.win32 && Config.ccomp_type = "msvc")
      then ["\"/EHsc\""]
-     else ["-Wall -Wextra -Wno-unused-parameter -x c++"]) @
+     else ["-Wall -Wextra -Wno-unused-parameter -x c++ -std=gnu++11"]) @
     (ifc useGLPK ["-DUSEGLPK"]) @
     (ifc useCOIN ["-DUSECOIN"]) @
     (ifc useCLP  ["-DUSECLP"]) @

